plugins/uefi/fu-plugin-uefi.c

```diff
@@ -613,9 +612,63 @@ load_custom_esp (FuPlugin *plugin, GError **error)
 
 			return FALSE;
 		}
+#ifdef HAVE_FWUP_CUSTOM_ESP
+		fwup_set_esp_mountpoint (data->esp_path);
 		g_debug ("%s set to %s", key, data->esp_path);
-		fu_plugin_add_report_metadata (plugin, key,
-					       data->esp_path);
+#endif
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_uefi_delete_old_capsules (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	const gchar *pattern = "/boot/efi/EFI/*/fw/fwupdate-*.cap";
+	g_autoptr(GPtrArray) files = NULL;
+
+	/* delete any files matching the glob in the ESP */
+	files = fu_common_get_files_recursive (data->esp_path, error);
+	if (files == NULL)
+		return FALSE;
+	for (guint i = 0; i < files->len; i++) {
+		const gchar *fn = g_ptr_array_index (files, i);
+		if (fnmatch (pattern, fn, 0) == 0) {
+			g_autoptr(GFile) file = g_file_new_for_path (fn);
+			g_debug ("deleting %s", fn);
+			if (!g_file_delete (file, NULL, error))
```

Comment:
It would be good to also delete the nvram variable not just the capsule. That will help prevent problems from folks who switch distros and keep stale variables around.

Reply:
Isn't that what the fwupdate cleanup oneshot does? I'd really like to keep this patch self-contained and as small as possible.

Reply:
Yeah except I've noticed it doesn't seem to be run automatically on all distros.  Arch or RHEL both have reports of old entries that weren't cleaned at fwupdate/fwupd installation.
So if you're cleaning old entries may as well clean those too.

Reply:
Have you got an example of what an "old entry" might look like? I'm hesitant to delete the detailed fwupdate debugging in the efivars for instance.

Reply:
"Old" entry and "new" entry would look identical.  If you parse `/sys/firmware/efi/efivars/fwupdate-$(ESRT)-0abba7dc-e516-4167-bbf5-4d9d1c739416` using efivar it will point to the name of the file in the ESP.  That's how the EFI application finds the file on the ESP too.

Now the `FWUPDATE_VERBOSE` and `FWUPDATE_DEBUG_LOG` variables arewhat stores that debugging, so those you would not delete.

Reply:
> it will point to the name of the file in the ESP

Right, but as I understand it that's going to be in UTF-16? Not insurmountable, but if I understand what you're asking (only delete the .cap files we actually scheduled on this hardware) I'm wondering why we can't just delete all of them? The same with the efivars, can't we just delete anything we find in `/sys/firmware/efi/efivars/fwupdate-*`?

Reply:
Probably, as long as you have confidence that the attempt happened since fwupd was last started.

Reply:
Could we delete any file with a mtime less than the boot time of the machine?

Reply:
That seems reasonable.  ESP's are FAT32.  I know there are some mtime limitations, but they might only be around time accuracy (in the order of seconds).

Reply:
That works well for files on the ESP. The efivars virtual filesystem doesn't store the mtime. I think we can do something you suggested on another "thread" -- check if the BootNext is set. That'll only be set when we've scheduled a firmware update in practice.

Reply:
OK that approach seems fine to me.

Reply:
Done, using libefivar this time.